fleetctl --tunnel with invalid IP times out and panics

```
$ fleetctl --tunnel 10.10.10.10 list-units

panic: dial tcp 10.10.10.10:22: operation timed out

goroutine 1 [running]:
runtime.panic(0x2eab60, 0xc21000a780)
    /usr/local/go/src/pkg/runtime/panic.c:266 +0xb6
main.getRegistry(0xc210060ee0, 0xc210037820)
    /Users/robszumski/Documents/fleet/src/github.com/coreos/fleet/fleetctl/cmd.go:50 +0x1ab
main.listUnitsAction(0xc210060ee0)
    /Users/robszumski/Documents/fleet/src/github.com/coreos/fleet/fleetctl/list_units.go:32 +0x35
github.com/coreos/fleet/third_party/github.com/codegangsta/cli.Command.Run(0x34d7b0, 0xa, 0x0, 0x0, 0x386e70, ...)
    /Users/robszumski/Documents/fleet/src/github.com/coreos/fleet/third_party/github.com/codegangsta/cli/command.go:73 +0x994
github.com/coreos/fleet/third_party/github.com/codegangsta/cli.(*App).Run(0xc21007b000, 0xc21000a000, 0x4, 0x4, 0x7, ...)
    /Users/robszumski/Documents/fleet/src/github.com/coreos/fleet/third_party/github.com/codegangsta/cli/app.go:111 +0x855
main.main()
    /Users/robszumski/Documents/fleet/src/github.com/coreos/fleet/fleetctl/cmd.go:97 +0x9ae

goroutine 3 [chan receive]:
github.com/coreos/fleet/third_party/github.com/golang/glog.(*loggingT).flushDaemon(0x6e7000)
    /Users/robszumski/Documents/fleet/src/github.com/coreos/fleet/third_party/github.com/golang/glog/glog.go:839 +0x50
created by github.com/coreos/fleet/third_party/github.com/golang/glog.init·1
    /Users/robszumski/Documents/fleet/src/github.com/coreos/fleet/third_party/github.com/golang/glog/glog.go:406 +0x276

goroutine 4 [syscall]:
runtime.goexit()
    /usr/local/go/src/pkg/runtime/proc.c:1394
```

--- a/fleetctl/journal.go
+++ b/fleetctl/journal.go
@@ -55,7 +55,7 @@ func journalAction(c *cli.Context) {
 		sshClient, err = ssh.NewSSHClient("core", addr)
 	}
 	if err != nil {
-		log.Fatalf("Unable to establish SSH connection: %v", err)
+		log.Fatal(err.Error())
 	}
 
 	defer sshClient.Close()

--- a/fleetctl/ssh.go
+++ b/fleetctl/ssh.go
@@ -81,7 +81,7 @@ func sshAction(c *cli.Context) {
 		sshClient, err = ssh.NewSSHClient("core", addr)
 	}
 	if err != nil {
-		log.Fatalf("Unable to establish SSH connection: %v", err)
+		log.Fatal(err.Error())
 		return
 	}
 

--- a/fleetctl/status.go
+++ b/fleetctl/status.go
@@ -62,7 +62,7 @@ func printUnitStatus(c *cli.Context, r *registry.Registry, jobName string) {
 		sshClient, err = ssh.NewSSHClient("core", addr)
 	}
 	if err != nil {
-		log.Fatalf("Unable to establish SSH connection: %v", err)
+		log.Fatal(err.Error())
 	}
 
 	defer sshClient.Close()

--- a/ssh/ssh.go
+++ b/ssh/ssh.go
@@ -5,6 +5,7 @@ import (
 	"errors"
 	"net"
 	"os"
+	"time"
 
 	gossh "github.com/coreos/fleet/third_party/code.google.com/p/go.crypto/ssh"
 	"github.com/coreos/fleet/third_party/code.google.com/p/go.crypto/ssh/terminal"
@@ -33,9 +34,9 @@ func Shell(client *gossh.ClientConn) error {
 	defer session.Close()
 
 	modes := gossh.TerminalModes{
-		gossh.ECHO:		1,	// enable echoing
-		gossh.TTY_OP_ISPEED:	14400,	// input speed = 14.4kbaud
-		gossh.TTY_OP_OSPEED:	14400,	// output speed = 14.4kbaud
+		gossh.ECHO:          1,     // enable echoing
+		gossh.TTY_OP_ISPEED: 14400, // input speed = 14.4kbaud
+		gossh.TTY_OP_OSPEED: 14400, // output speed = 14.4kbaud
 	}
 
 	fd := int(os.Stdin.Fd())
@@ -90,7 +91,14 @@ func NewSSHClient(user, addr string) (*gossh.ClientConn, error) {
 		return nil, err
 	}
 
-	return gossh.Dial("tcp", addr, clientConfig)
+	var client *gossh.ClientConn
+	dialFunc := func(echan chan error) {
+		var err error
+		client, err = gossh.Dial("tcp", addr, clientConfig)
+		echan <- err
+	}
+	err = timeoutSSHDial(dialFunc)
+	return client, err
 }
 
 func NewTunnelledSSHClient(user, tunaddr, tgtaddr string) (*gossh.ClientConn, error) {
@@ -99,12 +107,24 @@ func NewTunnelledSSHClient(user, tunaddr, tgtaddr string) (*gossh.ClientConn, er
 		return nil, err
 	}
 
-	tunnelClient, err := gossh.Dial("tcp", tunaddr, clientConfig)
+	var tunnelClient *gossh.ClientConn
+	dialFunc := func(echan chan error) {
+		var err error
+		tunnelClient, err = gossh.Dial("tcp", tunaddr, clientConfig)
+		echan <- err
+	}
+	err = timeoutSSHDial(dialFunc)
 	if err != nil {
 		return nil, err
 	}
 
-	targetConn, err := tunnelClient.Dial("tcp", tgtaddr)
+	var targetConn net.Conn
+	dialFunc = func(echan chan error) {
+		var err error
+		targetConn, err = tunnelClient.Dial("tcp", tgtaddr)
+		echan <- err
+	}
+	err = timeoutSSHDial(dialFunc)
 	if err != nil {
 		return nil, err
 	}
@@ -115,4 +135,18 @@ func NewTunnelledSSHClient(user, tunaddr, tgtaddr string) (*gossh.ClientConn, er
 	}
 
 	return targetClient, nil
+}
+
+func timeoutSSHDial(dial func(chan error)) error {
+	var echan chan error
+	var err error
+
+	go dial(echan)
+
+	select {
+	case <-time.After(time.Duration(time.Second * 10)):
+		return errors.New("Timed out while initiating SSH connection")
+	case err = <-echan:
+		return err
+	}
 }

